src/hotspot/share/utilities/intpow.hpp

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2024, Arm Limited. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_INTPOW_HPP
+#define SHARE_UTILITIES_INTPOW_HPP
+
+#include "metaprogramming/enableIf.hpp"
+#include <limits>
+#include <type_traits>
+
+template <typename T, T v, unsigned p, bool no_overflow = false,
+          typename = std::enable_if_t<std::is_integral<T>::value && std::is_unsigned<T>::value>>
+struct intpow {
+  static_assert(v || p, "0^0 is not defined");
+
+ private:
+  // We use exponentiation by squaring to calculate the required power.
+  static const T _a = intpow<T, v, p / 2, no_overflow>::value;
+  static const T _b = (p % 2) ? v : 1;
+
+  static_assert(!no_overflow || _a <= std::numeric_limits<T>::max() / _a, "Integer overflow");
+  static_assert(!no_overflow || _a * _a <= std::numeric_limits<T>::max() / _b, "Integer overflow");
+
+ public:
+  static const T value = _a * _a * _b;
+};
+
+template <typename T, T v, bool no_overflow>
+struct intpow<T, v, 0, no_overflow> {
+  static const T value = 1;
+};
+
+template <typename T, T v, bool no_overflow>
+struct intpow<T, v, 1, no_overflow> {
+  static const T value = v;
+};
```

Comment:
> There's no need for any of this metaprogramming. A constexpr function would be better.

The main advantage of this implementation over one using a `constexpr` function is the ability to verify input parameter values and detect overflows at compile time using `static_assert`. This feature helps prevent user errors that could otherwise be difficult to debug. I chose this approach for its functional benefits over a `constexpr` function. If you have any suggestions on how to implement this functionality in a `constexpr` function, they would be highly welcome, as I might be missing something here. :smiley:

Below is a similar `constexpr` implementation. **Please don't accept it through the UI**. I kept a `typedef` template parameter to ensure the return value has the desired width, if needed, without applying a mask at the call site.

If you'd still prefer a `constexpr` function, please let me know if you have any comments on the implementation below (I'll rename it, remove `no_overflow` parameter and the comments).

```suggestion
#include <limits>

template <typename T>
static constexpr T intpow_c(T v, unsigned p, bool no_overflow = false) {
  // Can't be used in constexpr function
  // static_assert(v || p, "0^0 is not defined");

  if (p == 0) {
    return 1;
  }

  T a = intpow_c(v, p / 2, no_overflow);
  T b = (p % 2) ? v : 1;

  // Can't be used in constexpr function
  // static_assert(!no_overflow || a <= std::numeric_limits<T>::max() / a, "Integer overflow");
  // static_assert(!no_overflow || a * a <= std::numeric_limits<T>::max() / b, "Integer overflow");

  return a * a * b;
}
```

Reply:
You're assuming a requirement that isn't really there. HotSpot calculates a great many things at startup time, and given that the arguments to intpow here are constants, there's  no possibility of any surprises at runtime. Just assert what needs to be asserted, and please keep utility functions like this as simple as possible, leaving template metaprogramming for places that need it.

Reply:
So yes, this looks OK, but of course the "Can't use" comments are unnecessary if you just use` assert`s.

P.S. If T is unsigned, there is no overflow. The desired result is mod 2**n, by the definition of the unsigned integer types. 

P.P.S 0**0 is 1, for all purposes for which integer pow() is going to be used in HotSpot: this is combinatorics rather than calculus.  It's also what `Math.pow()` does, I think so we're good.

Reply:
> P.S. If T is unsigned, there is no overflow. The desired result is mod 2**n, by the definition of the unsigned integer types.

The idea behind `no_overflow` was to allow user to verify that the result is the exact representation of `v**p` and not just `mod 2**n`. Since this is not required, I'll remove the parameter and the assertions and add `std::is_unsigned<T>::value` constraint instead.